plutus-chain-index/app/Marconi.hs

```diff
@@ -174,50 +156,106 @@ getUtxoUpdate slot txs =
                  , _slotNo  = slot
                  }
 
-utxoIndexer
-  :: FilePath
+data Coordinator = Coordinator
+  { _channel      :: TChan (ChainSyncEvent (BlockInMode CardanoMode))
+  , _barrier      :: QSemN
+  , _indexerCount :: Int
+  }
+
+initialCoordinator :: Int -> IO Coordinator
+initialCoordinator indexerCount =
+  Coordinator <$> newBroadcastTChanIO
+              <*> newQSemN 0
+              <*> pure indexerCount
+
+datumWorker
+  :: Coordinator
+  -> TChan (ChainSyncEvent (BlockInMode CardanoMode))
+  -> FilePath
+  -> IO ()
+datumWorker Coordinator{_barrier} ch path = Datum.open path (Datum.Depth 2160) >>= innerLoop
+  where
+    innerLoop :: DatumIndex -> IO ()
+    innerLoop index = do
+      signalQSemN _barrier 1
+      event <- atomically $ readTChan ch
+      case event of
+        RollForward blk _ct ->
+          Ix.insert (getDatums blk) index >>= innerLoop
+        RollBackward cp _ct -> do
+          events <- Ix.getEvents (index ^. Ix.storage)
+          innerLoop $
+            fromMaybe index $ do
+              slot   <- chainPointToSlotNo cp
+              offset <- findIndex (any (\(s, _) -> s < slot)) events
+              Ix.rewind offset index
+
+utxoWorker
+  :: Coordinator
+  -> TChan (ChainSyncEvent (BlockInMode CardanoMode))
+  -> FilePath
+  -> IO ()
+utxoWorker Coordinator{_barrier} ch path = Utxo.open path (Utxo.Depth 2160) >>= innerLoop
+  where
+    innerLoop :: UtxoIndex -> IO ()
+    innerLoop index = do
+      signalQSemN _barrier 1
+      event <- atomically $ readTChan ch
+      case event of
+        RollForward (BlockInMode (Block (BlockHeader slotNo _ _) txs) _) _ct ->
+          Ix.insert (getUtxoUpdate slotNo txs) index >>= innerLoop
+        RollBackward cp _ct -> do
+          events <- Ix.getEvents (index ^. Ix.storage)
+          innerLoop $
+            fromMaybe index $ do
+              slot   <- chainPointToSlotNo cp
+              offset <- findIndex  (\u -> (u ^. Utxo.slotNo) < slot) events
+              Ix.rewind offset index
+
+combinedIndexer
+  :: Maybe FilePath
+  -> Maybe FilePath
```

Comment:
This approach does not seem extensible for multiple indexers. Why not pass a list of workers? They seem to have a similar interface.

Reply:
It can probably be done, but it's not trivial. You need to provide specific configurations for each indexer, and the startup may also be different (not the case here). The issue mentioned 2 indexers and I am not sure if supporting multiple indexers is on the bird path, but if it is, it should be a separate issue, I think.

Reply:
:+1: